### ISSUE TITLE
[sled-agent] Failing to start the guest leaves instance state as `starting`

### ISSUE THREAD

**Issue by bnaecker** [oxidecomputer/omicron#831]
I'm creating a guest instance using the Oxide CLI. That provision failed, but we can still fetch the database record for the guest, which shows that it's still in the `"starting"` state.

```
bnaecker@feldspar : ~/cli $ cargo run -- instance create -p my-proj -o my-org vm
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `target/debug/oxide instance create -p my-proj -o my-org vm`
instance description:: vm
instance hostname:: vm
instance memory:: 4
instance ncpus:: 2
✘ Oxide API internal error: Internal Server Error
bnaecker@feldspar : ~/cli $ cargo run -- instance get -p my-proj -o my-org vm
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
     Running `target/debug/oxide instance get -p my-proj -o my-org vm`
 time_run_state_updated | 22 seconds ago
 time_modified          | 22 seconds ago
 time_created           | 22 seconds ago
 run_state              | creating
 project_id             | 78ab867f-6b1e-440d-af31-c336c7872e50
 ncpus                  | 2
 memory                 | 4
 hostname               | vm
 description            | vm
 name                   | vm
 id                     | 891deb61-aed8-4a69-82bc-f243633076e5

bnaecker@feldspar : ~/cli $
```

The sled agent logs show:

```
Mar 29 20:17:42.237 INFO accepted connection, remote_addr: 192.168.1.123:44314, local_addr: 192.168.1.201:12345, component: dropshot
Mar 29 20:17:42.238 INFO instance_ensure 891deb61-aed8-4a69-82bc-f243633076e5 -> InstanceRuntimeStateRequested { run_state: Running, migration_params: None }, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:42.238 INFO new instance, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:42.238 INFO Instance::new w/initial HW: InstanceHardware { runtime: InstanceRuntimeState { run_state: Creating, sled_uuid: fb0f7546-4d46-40ca-9d56-cbb810684ca7, propolis_uuid: 53f196cb-b5cc-4ce6-bb6c-e299ccd4c7ea, dst_propolis_uuid: None, propolis_addr: None, migration_uuid: None, ncpus: InstanceCpuCount(2), memory: ByteCount(4), hostname: "vm", gen: Generation(1), time_updated: 2022-03-29T20:17:42.181121Z }, nics: [NetworkInterface { name: Name("default"), ip: 172.30.0.13, mac: MacAddr(MacAddr6([168, 64, 37, 255, 19, 4])), subnet: V4(Ipv4Net(Ipv4Network { addr: 172.30.0.0, prefix: 22 })), vpc_id: 00dc226d-6198-415b-b6a2-6a1a91cfb63f }] }, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:42.239 INFO created OPTE port for guest, port_info: OptePort { name: "opte1", ip: 172.30.0.13, subnet: V4(Ipv4Network { addr: 172.30.0.0, prefix: 22 }), mac: MacAddr6([168, 64, 37, 255, 19, 4]), vni: Vni { inner: 56354 }, underlay_ip: fd00::1, gateway: Gateway { mac: MacAddr6([168, 64, 37, 0, 0, 1]), ip: 172.30.0.1 }, boundary_services: BoundaryServices { ip: fd00:99::1, vni: Vni { inner: 99 } } }, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:42.274 INFO Configuring new Omicron zone: oxz_propolis-server_53f196cb-b5cc-4ce6-bb6c-e299ccd4c7ea, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:42.297 INFO Installing Omicron zone: oxz_propolis-server_53f196cb-b5cc-4ce6-bb6c-e299ccd4c7ea, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:44.749 INFO Zone booting, zone: oxz_propolis-server_53f196cb-b5cc-4ce6-bb6c-e299ccd4c7ea, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:46.698 INFO Adding address: Dhcp, zone: oxz_propolis-server_53f196cb-b5cc-4ce6-bb6c-e299ccd4c7ea, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:53.130 INFO Created address 192.168.1.205/24 for zone: oxz_propolis-server_53f196cb-b5cc-4ce6-bb6c-e299ccd4c7ea, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:53.582 INFO Started propolis in zone: oxz_propolis-server_53f196cb-b5cc-4ce6-bb6c-e299ccd4c7ea, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:53.899 INFO GET request to http://192.168.1.205:12400/instances/00000000-0000-0000-0000-000000000000, propolis_client address: 192.168.1.205:12400, component: propolis-client, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:53.900 INFO Sending ensure request to propolis: InstanceEnsureRequest { properties: InstanceProperties { id: 891deb61-aed8-4a69-82bc-f243633076e5, name: "vm", description: "Test description", image_id: 00000000-0000-0000-0000-000000000000, bootrom_id: 00000000-0000-0000-0000-000000000000, memory: 0, vcpus: 2 }, nics: [NetworkInterfaceRequest { name: "oxGuestInstance7", slot: Slot(0) }], disks: [], migrate: None }, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:53.900 INFO PUT request to http://192.168.1.205:12400/instances/891deb61-aed8-4a69-82bc-f243633076e5, propolis_client address: 192.168.1.205:12400, component: propolis-client, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:53.910 INFO halt_and_remove: Zone state: Running, zone: oxz_propolis-server_53f196cb-b5cc-4ce6-bb6c-e299ccd4c7ea, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:55.824 INFO Stopped and uninstalled zone, zone: oxz_propolis-server_53f196cb-b5cc-4ce6-bb6c-e299ccd4c7ea, instance_id: 891deb61-aed8-4a69-82bc-f243633076e5, sled_id: fb0f7546-4d46-40ca-9d56-cbb810684ca7, component: SledAgent
Mar 29 20:17:55.842 INFO request completed, error_message_external: Internal Server Error, error_message_internal: Error managing instances: Instance error: Failure from Propolis Client: Bad Status: 500, response_code: 500, uri: /instances/891deb61-aed8-4a69-82bc-f243633076e5, method: PUT, req_id: 8e30213e-18bd-4387-9ed1-0f1d7fe8fc92, remote_addr: 192.168.1.123:44314, local_addr: 192.168.1.201:12345, component: dropshot
```

Note that this is on the current `opteadm-integration` branch, not `main`, though the failure should be handled the same in any case. Nexus should handle this failure by updating the instance record to the `failed` state, or possibly remove the record entirely. I'm not sure which of those two is more appropriate, but the latter seems to map more directly to the underlying saga.

**Comment by bnaecker** [comment]
So, this isn't really the sled agent's fault. Nexus still has a no-op undo action in the instance-creation saga. That should be changed to stop and then delete the instance. The former is just needed because we can't delete things in the "starting" state.

**Comment by smklein** [comment]
> So, this isn't really the responsibility of the sled agent. Nexus still has a no-op undo action in the instance-creation saga. That should be changed to stop and then delete the instance. The former is just needed because we can't delete things in the "starting" state.

After the instance has been created, Sled Agent is *normally* in charge of reporting state changes to Nexus:

https://github.com/oxidecomputer/omicron/blob/0d19918e44d440cf59b0757a62080e60989027f6/sled-agent/src/instance.rs#L225-L234

However, in this case, I agree with you - we only start monitoring the state in Sled Agent *after* the call to `instance_ensure` in propolis was successful. In this case, when it wasn't, Nexus needs to know to mark the instance record as failed.

**Comment by bnaecker** [comment]
This was actually closed by #806, I had forgotten to rebase before I made this. Dropping.
